cmd/cluster-init/generate_job.go

```diff
@@ -0,0 +1,182 @@
+			Agent:      string(prowapi.KubernetesAgent),
+			Cluster:    string(api.ClusterAPPCI),
+			SourcePath: "",
+			Spec: &v1.PodSpec{
+				Volumes:            []v1.Volume{generateSecretVolume(clusterName)},
+				Containers:         []v1.Container{generateContainer(LatestImage, args, generateVolumeMounts())},
+				ServiceAccountName: ConfigUpdater,
+			},
+			UtilityConfig: prowconfig.UtilityConfig{
+				Decorate: &trueBool,
+			},
+			MaxConcurrency: 1,
+			Labels: map[string]string{
+				OpenshiftRole: Infra,
+			},
+		},
+		Brancher: prowconfig.Brancher{
+			Branches: []string{"master"},
+		},
+	}
+}
+
+func generatePresubmit(clusterName string) prowconfig.Presubmit {
+	mounts := generateVolumeMounts()
+	mounts = append(mounts, v1.VolumeMount{
+		Name:      Tmp,
+		MountPath: "/" + Tmp,
+	})
+	return prowconfig.Presubmit{
+		JobBase: prowconfig.JobBase{
+			Name:       "pull-ci-openshift-release-master-" + clusterName + "-dry",
+			Agent:      string(prowapi.KubernetesAgent),
+			Cluster:    string(api.ClusterAPPCI),
+			SourcePath: "",
+			Spec: &v1.PodSpec{
+				Volumes: []v1.Volume{generateSecretVolume(clusterName),
+					{
+						Name: Tmp,
+						VolumeSource: v1.VolumeSource{
+							EmptyDir: &v1.EmptyDirVolumeSource{},
+						},
+					}},
+				Containers:         []v1.Container{generateContainer(LatestImage, generateArgs(clusterName), mounts)},
+				ServiceAccountName: ConfigUpdater,
+			},
+			UtilityConfig: generateUtilityConfig(),
+			Labels: map[string]string{
+				"pj-rehearse.openshift.io/can-be-rehearsed": "true",
+			},
+		},
+		AlwaysRun:    true,
+		Optional:     false,
+		Trigger:      prowconfig.DefaultTriggerFor(clusterName),
+		RerunCommand: prowconfig.DefaultRerunCommandFor(clusterName) + "-dry",
+		Brancher: prowconfig.Brancher{
+			Branches: []string{"master"},
+		},
+		Reporter: prowconfig.Reporter{
+			Context: fmt.Sprintf("ci/build-farm/%s-dry", clusterName),
+		},
+	}
+}
+
+func generateUtilityConfig() prowconfig.UtilityConfig {
+	trueBool := true
+	return prowconfig.UtilityConfig{
+		Decorate:  &trueBool,
+		ExtraRefs: []prowapi.Refs{},
+	}
+}
+
+func generateSecretVolume(clusterName string) v1.Volume {
+	return v1.Volume{
+		Name: "build-farm-credentials",
+		VolumeSource: v1.VolumeSource{
+			Secret: &v1.SecretVolumeSource{
+				SecretName: "build-farm-credentials",
+				Items: []v1.KeyToPath{
+					{
+						Key:  serviceAccountKubeconfigPath(ConfigUpdater, clusterName),
+						Path: "kubeconfig",
+					},
+				},
+			},
+		},
+	}
+}
+
+func generateContainer(image string, args []string, volumeMounts []v1.VolumeMount) v1.Container {
+	return v1.Container{
+		Name:    "",
+		Image:   image,
+		Command: []string{"applyconfig"},
+		Args:    args,
+		Resources: v1.ResourceRequirements{
+			Requests: map[v1.ResourceName]resource.Quantity{
+				"cpu": resource.MustParse("10m"),
+			},
+		},
+		ImagePullPolicy: "Always",
+		VolumeMounts:    volumeMounts,
+	}
+}
+
+func generateArgs(clusterName string) []string {
```

Comment:
For baseline args, volume mounts, etc - if they are *always* called and *always* inputs to `generateContainer`, have them in there inline and allow the user to pass _additional_ args that get appended.

Reply:
That makes sense. I'm still not completely sold on all of the inlining, but I'm starting to get more used to it.